napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,476 @@
+"""Plugin related functions that require Qt.
+
+Non-Qt plugin functions can be found in: `napari/plugins/_npe2.py`
+"""
+
+from __future__ import annotations
+
+import inspect
+from functools import partial
+from itertools import chain
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from app_model import Action
+from app_model.types import SubmenuItem, ToggleRule
+from magicgui.type_map._magicgui import MagicFactory
+from magicgui.widgets import FunctionGui, Widget
+from npe2 import plugin_manager as pm
+from qtpy.QtWidgets import QWidget
+
+from napari._app_model import get_app
+from napari._app_model.constants import MenuGroup, MenuId
+from napari._app_model.injection._providers import _provide_viewer
+from napari._qt._qapp_model.injection._qproviders import _provide_window
+from napari.errors.reader_errors import MultipleReaderError
+from napari.plugins import menu_item_template, plugin_manager
+from napari.plugins._npe2 import get_widget_contribution
+from napari.utils.events import Event
+from napari.utils.translations import trans
+from napari.viewer import Viewer
+
+if TYPE_CHECKING:
+    from npe2.manifest import PluginManifest
+    from npe2.plugin_manager import PluginName
+    from npe2.types import WidgetCreator
+
+    from napari.qt import QtViewer
+
+
+# TODO: This is a separate function from `_build_samples_submenu_actions` so it
+# can be easily deleted once npe1 is no longer supported.
+def _rebuild_npe1_samples_menu() -> None:
+    """Register submenu and actions for all npe1 plugins, clearing all first."""
+    app = get_app()
+    # Unregister all existing npe1 sample menu actions and submenus
+    if unreg := plugin_manager._unreg_sample_submenus:
+        unreg()
+    if unreg := plugin_manager._unreg_sample_actions:
+        unreg()
+
+    sample_actions: List[Action] = []
+    for plugin_name, samples in plugin_manager._sample_data.items():
+        multiprovider = len(samples) > 1
+        if multiprovider:
+            submenu_id = f'napari/file/samples/{plugin_name}'
+            submenu = [
+                (
+                    MenuId.FILE_SAMPLES,
+                    SubmenuItem(
+                        submenu=submenu_id, title=trans._(plugin_name)
+                    ),
+                ),
+            ]
+        else:
+            submenu_id = MenuId.FILE_SAMPLES
+            submenu = []
+
+        for sample_name, sample_dict in samples.items():
+
+            def _add_sample(
+                qt_viewer: QtViewer,
+                plugin: str = plugin_name,
+                sample: str = sample_name,
+            ) -> None:
```

Comment:
We should never define function inside for loop body. It will depend on function namespace so, when it will be called it will use `sample_name` and `sample_dict` from last loop

Reply:
Yes agreed, I amended the npe2 one to use a partial (separate PR) but not this one. I think I wanted to do in separate PR as this one is already so big. This PR does move the code (but does not change it) - but I can do it here if you wish. Or a separate PR.